### PR TITLE
Feature/migrate-spring-data-cosmos

### PR DESCRIPTION
**Pull request by yiliuTo**
Migrate source code and unit tests from repo spring-data-cosmos here.

Change the dependency in azure-spring-boot of spring-data-cosmos from external to current.

Keep several dependency versions different from SDK including
- spring_org.springframework:spring-web
- spring_com.microsoft.azure:azure-cosmos
- spring_org.springframework.boot:spring-boot-starter-test

### CODE REVIEW

File: sdk/spring/azure-spring-boot/pom.xml [hunk 1]
```diff
@@ -251,6 +251,12 @@
       <version>1.1.1</version> <!-- {x-version-update;pl.pragmatists:JUnitParams;external_dependency} -->
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>com.azure</groupId>
+      <artifactId>azure-identity</artifactId>
+      <version>1.0.4</version>
```
Comment (alzimmermsft) [inline sdk/spring/azure-spring-boot/pom.xml:257]: Shouldn't this be using the latest GA version available which is `1.0.7`?

File: sdk/spring/azure-spring-data-cosmosdb/QueryPartitionedCollection.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        String street;
+        @PartitionKey
+        String city;
+    }
+```
+
+Write the repository interface:
+```
+    @Repository
+    public interface AddressRepository extends DocumentDbRepository<Address, String> {
```
Comment (kushagraThapar) [inline sdk/spring/azure-spring-data-cosmosdb/QueryPartitionedCollection.md:26]: There is no `DocumentDbRepository`. That is in old spring-data-cosmosdb SDK. 
This should extend `CosmosRepository` or `ReactiveCosmosRepository`

File: sdk/spring/azure-spring-data-cosmosdb/QueryPartitionedCollection.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
+### How to Query Partitioned Azure Cosmos DB Collection
```
Comment (kushagraThapar) [inline sdk/spring/azure-spring-data-cosmosdb/QueryPartitionedCollection.md:1]: What is the use of this file ?
This should be merged with README file, or move this topic to spring-data-cosmosdb developer's guide : https://docs.microsoft.com/en-us/azure/developer/java/spring-framework/how-to-guides-spring-data-cosmosdb

File: sdk/spring/azure-spring-boot/pom.xml (hunk 1 above)
Comment (yiliuTo) [reply sdk/spring/azure-spring-boot/pom.xml:257]: Yes, this snippet didn't be modified by rebasing and I will delete it manually.

File: sdk/spring/azure-spring-data-cosmosdb/QueryPartitionedCollection.md (hunk 3 above)
Comment (yiliuTo) [reply sdk/spring/azure-spring-data-cosmosdb/QueryPartitionedCollection.md]: The content of this file has been included into README file about using partitioned query, thus I will remove this file.

File: eng/versioning/external_dependencies.txt [hunk 4]
```diff
@@ -251,3 +263,9 @@ storage_com.microsoft.azure:azure-storage;8.4.0
 spring_io.micrometer:micrometer-core;1.3.0
 spring_io.micrometer:micrometer-registry-azure-monitor;1.3.0
 spring_com.microsoft.azure:azure;1.34.0
+
+#sdk\spring\spring-data-cosmosdb\pom.xml
+spring_org.springframework:spring-web;5.2.6.RELEASE
+spring_com.microsoft.azure:azure-cosmos;3.7.3
+spring_org.springframework.boot:spring-boot-starter-test;2.3.0.RELEASE
```
[the remark this answers is not in the document]
Comment (yiliuTo) [reply eng/versioning/external_dependencies.txt]: Thanks for the suggestions. I have rebased the related PR and updated version of the dependencies.

**Comment by yiliuTo** [comment]
Hi @JimSuplizio, due to the -amd parameter in java-core-ci pipeline, the azure-spring-boot project gets triggered while spring-data-cosmosdb not, however azure-spring-boot depends on spring-data-cosmosdb thus the java-core-ci pipeline always fails.
To deal with this problem, currently I add a useless dependency azure-identity in spring-data-cosmos and now the pipeline can pass now. It seems such solution doesn't solve the problem completely and imports a useless dependency. Could you please help us with some suggestions? Thank you~

**Comment by yiliuTo** [comment]
In order to pass the java-core-ci pipeline, currently I add the spring-data-cosmosdb as the _external dependency_ for azure-spring-boot project given that it cannot be triggered out as a _dependency_ of track 2.

**Comment by saragluna** [comment]
Hi @JimSuplizio and @kushagraThapar, could you please take a look at this PR?

**Comment by JimSuplizio** [comment]
> In order to pass the java-core-ci pipeline, currently I add the spring-data-cosmosdb as the _external dependency_ for azure-spring-boot project given that it cannot be built as a _dependency_ of track 2.

This is exactly what I'd said [needed to get done] (https://github.com/Azure/azure-sdk/issues/892#issuecomment-646853728). Until the spring-data-cosmosdb, being added to the repository, is updated to target v4 cosmos, spring is going to have to use the released versions as external dependencies.

File: eng/versioning/version_client.txt [hunk 5]
```diff
@@ -55,7 +55,7 @@ com.microsoft.azure:azure-servicebus-jms-spring-boot-starter;2.3.2;2.3.3-beta.1
 com.microsoft.azure:azure-spring-boot-metrics-starter;2.3.2;2.3.3-beta.1
 com.microsoft.azure:azure-spring-boot-tests;2.3.2;2.3.3-beta.1
 com.microsoft.azure:azure-spring-boot-test-core;2.3.2;2.3.3-beta.1
-
+com.microsoft.azure:spring-data-cosmosdb;2.3.0;2.3.1-beta.1
```
Comment (JimSuplizio) [inline eng/versioning/version_client.txt:58]: @kushagraThapar, as per our discussion earlier we'll let this get checked in like this but part of the v4 updates will be to rename the groupId and change the artifactId to match the naming convention. Also, the work item you're creating will also encompass updating the versions to match this update to v4 cosmos which will update the major to 3.

Comment (kushagraThapar) [reply eng/versioning/version_client.txt:58]: @JimSuplizio - Sounds good. Here is the issue : https://github.com/Azure/azure-sdk-for-java/issues/12490

**Pull request merged**